src/cryptonote_basic/miner.cpp

```diff
@@ -846,40 +846,109 @@ namespace cryptonote
-        "/sys/class/power_supply/ADP0/online"
-      };
+      // Use the power_supply class http://lxr.linux.no/#linux+v4.10.1/Documentation/power/power_supply_class.txt
+      std::string power_supply_class_path = "/sys/class/power_supply";
 
-      for(const std::string& path : POWER_SUPPLY_STATUS_PATHS)
+      if (epee::file_io_utils::is_directory(power_supply_class_path))
       {
-        if( epee::file_io_utils::is_file_exist(path) )
-        {
-          power_supply_path = path;
-          break;
-        }
-      }
+        std::vector<boost::filesystem::path> paths;
+        epee::file_io_utils::get_directory_paths(power_supply_class_path, paths);
 
-      if( power_supply_path.empty() )
-      {
-        LOG_ERROR("Couldn't find battery/power status file, can't determine if plugged in!");
-        return boost::logic::tribool(boost::logic::indeterminate);;
-      }
+        // Iterate through power_supplies alphabetically.
+        // This ensures that we iterate through potential
+        // AC power supplies first.
+        std::sort(paths.begin(), paths.end());
 
-      std::ifstream power_stream(power_supply_path);
-      if( power_stream.fail() )
-      {
-        LOG_ERROR("failed to open '" << power_supply_path << "'");
-        return boost::logic::tribool(boost::logic::indeterminate);;
+        // Take the first present power supply as truth
+        for (std::vector<boost::filesystem::path>::const_iterator iter(paths.begin()), iter_end(paths.end()); iter != iter_end; ++iter)
+        {
+          std::string power_supply_path = (*iter).string();
+          if (epee::file_io_utils::is_directory(power_supply_path))
+          {
+            std:string power_supply_present = epee::file_io_utils::file_to_string(power_supply_path + "/present");
+            if (!boost::starts_with(power_supply_present, "1")) {
+              LOG_PRINT_L4("Power supply not present at " << power_supply_path);
+              continue;
+            }
+
+            std::string power_supply_type_path = power_supply_path + "/type";
+            if (epee::file_io_utils::is_regular_file(power_supply_type_path))
+            {
+              std::string power_supply_type = epee::file_io_utils::file_to_string(power_supply_type_path);
+
+              if (boost::starts_with(power_supply_type, "Battery"))
+              {
+                std::string power_supply_status_path = power_supply_path + "/status";
+                if (epee::file_io_utils::is_regular_file(power_supply_status_path))
+                {
+                  std::string power_supply_status = epee::file_io_utils::file_to_string(power_supply_status_path);
+
+                  if (boost::starts_with(power_supply_status, "Charging"))
+                  {
+                    return boost::logic::tribool(false);
+                  }
+                  else if (boost::starts_with(power_supply_status, "Full"))
+                  {
+                    return boost::logic::tribool(false);
```

Comment:
> But it's very difficult going through a peer review with someone who doesn't know the spec or what acpi is.

Well that's a bit rich _and_ a big assumption. Or are you talking about yourself? The 'spec' doesn't define the filenames or define how BAT and AC should switch. Hence constructive comments to try and catch all cases to find a solution that catches all eventualities. If you know the acpi and spec better than the rest of us then your PR would have been better and had only positive reviews.